parsely/src/main/java/com/parsely/parselyandroid/SendEvents.kt

```diff
@@ -0,0 +1,57 @@
+package com.parsely.parselyandroid
+
+import com.parsely.parselyandroid.JsonSerializer.toParselyEventsPayload
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+
+internal class SendEvents(
+    private val flushManager: FlushManager,
+    private val localStorageRepository: LocalStorageRepository,
+    private val parselyAPIConnection: ParselyAPIConnection,
+    private val scope: CoroutineScope
+) {
+
+    private val mutex = Mutex()
+
+    operator fun invoke(isDebug: Boolean) {
+        scope.launch {
+            mutex.withLock {
+                val eventsToSend = localStorageRepository.getStoredQueue()
+
+                if (eventsToSend.isEmpty()) {
+                    flushManager.stop()
+                    return@launch
+                }
+                ParselyTracker.PLog("Sending request with %d events", eventsToSend.size)
+
+                val jsonPayload = toParselyEventsPayload(eventsToSend)
+
+                ParselyTracker.PLog("POST Data %s", jsonPayload)
+
+                if (isDebug) {
+                    ParselyTracker.PLog("Debug mode on. Not sending to Parse.ly")
+                    localStorageRepository.remove(eventsToSend)
+                } else {
+                    ParselyTracker.PLog("Requested %s", ParselyTracker.ROOT_URL)
+                    parselyAPIConnection.send(jsonPayload)
+                        .fold(
+                            onSuccess = {
+                                ParselyTracker.PLog("Pixel request success")
+                                localStorageRepository.remove(eventsToSend)
+                                ParselyTracker.PLog("Event queue empty, flush timer cleared.")
+                                if (localStorageRepository.getStoredQueue().isEmpty()) {
+                                    flushManager.stop()
+                                }
```

Comment:
Would it make sense to skip this part and let the `if (eventsToSend.isEmpty())` check handle this logic for us?

Reply:
I'm not sure: between `eventsToSend.isEmpty` check and sending events, the consumer app might report new events, which are saved to the stored queue. In such case, we don't want to stop the flush timer. That's why I added check for local storage just before deciding if we want to stop flush manager. 

Does it make sense, WDYT?